drivers: spi: Add basic support rpi-pico SPI (PrimeCell SSP PL022)

soburi opened the pull request:
Add fundamental feature support for rpi-pico's SPI peripheral.
This commit implements synchronous transfer with 8bit-MSB format.

I verified my environment with two test programs.
tests/drivers/spi/spi_loopback test is work well.
(rpi_pico seems to have a problem when running device-testing  by the twister,
 locally build and flashing and check console output.)
And also samples/subsys/fs/fat_fs/ works.

rp2040 datasheet says the SoC has ARM PrimeCell SSP PL022 as the SPI peripheral.
So this PL022 driver implementation can compile without rpi_pico SDK.

Optional feature are not currently implemented yet.
It is work in future.

- DMA transfer is not implemented yet.
- Slave mode is not implemented yet.
- Currently support only 8-bit data transfer.

Hardware limitation:

- LSB-first format is not supported by hardware.

soburi committed:
drivers: spi: Add basic support for PL022

Add fundamental feature support for PL022 SPI peripheral.
This commit implements synchronous transfer with 8bit-MSB format.

Optional functions are not currently implemented yet.

- interrupt based transfer is not implemented yet.
- DMA transfer is not implemented yet.
- Slave mode is not implemented yet.
- Currently support only 8-bit data transfer.

Hardware limitation:

- LSB-first format is not supported by hardware.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
dts: arm: rpi_pico: Adding spi configurations to rp2040

Add spi0 and spi1 configuration to rp2040.

spi0 and spi1 are declared as compatible with
"raspberrypi,pico-spi" and "arm,pl022".
"raspberrypi,pico-spi" is just inherited from
"arm,pl022" and "reset-device"

spi0 and spi1 declare as compatible with "raspberrypi,pico-spi"
and "arm,pl022".
"raspberrypi,pico-spi" is just inheriting "arm,pl022" and
"reset-device" for declaring the 'reset' property,
with no additional property and implementation.

Add also pinctrl macros to dt-bindings header.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: arm: rpi_pico: Adding spi configuration to rpi_pico

Add configuration into dts to enable spi0 and spi1.
In default, GP16 to GP19 using by spi0, and GP10 to GP13 using by spi1.
Also, update rpi_pico.yaml to indicate this board supporting spi.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

yonsch commented on boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi:
```diff
@@ -31,4 +31,18 @@
+
+	spi0_default: spi0_default {
+		group1 {
+			pinmux = <SPI0_RX_P16>, <SPI0_TX_P19>, <SPI0_SCK_P18>;
```
TX and SCK should be in their own group, where input is not enabled

soburi replied:
Fix it.

yonsch commented on boards/arm/rpi_pico/rpi_pico.yaml:
```diff
@@ -12,4 +12,5 @@ supported:
   - uart
   - gpio
   - i2c
+  - spi
```
Add SPI to the "Supported Features" table in `boards/arm/rpi_pico/doc.index.rst` also`

soburi replied:
Fixed it.

yonsch commented on dts/arm/rpi_pico/rp2040.dtsi:
```diff
@@ -100,6 +100,26 @@
 		};
 
+		spi0: spi@4003c000 {
+			compatible = "arm,pl022";
```
SPI should also be a reset device. @andrei-edward-popa , I suppose the correct way to do this is to create a separate DTS binding for rpi_pico, and have it inherit pl022 and reset-device?

andrei-edward-popa replied:
@soburi @yonsch Yes, you can create a DTS bindings under dts/bindings/spi/raspberrypi,pico-spi.yaml and inherit arm,pl022.yaml and reset,device.yaml (see raspberrypi,pico-uart.yaml). After that, add resets field inside spi nodes (take a look inside rpi_pico_common.dtsi for spi reset values).

yonsch replied:
Right, you don't need another implementation. Just make sure that enabling SPI in DTS still selects the PL022 implementation

andrei-edward-popa replied:
I think you can remove arm,pl022 from the compatible strings because raspberrypi,pico-spi yaml inherits arm,pl022.yaml.

soburi replied:
> I think you can remove arm,pl022 from the compatible string because raspberrypi,pico-spi yaml inherits arm,pl022.yaml.

If do so, It couldn't resolve pl022 in DT_INST_FOREACH_STATUS_OKAY, can't compile it.
It seem to need it.

Compiling with CONFIG_SPI_PL022, I get such as,

```
/home/crs/zephyrproject/zephyr/drivers/spi/spi_pl022.c:473:12: warning: 'spi_pl022_init' defined but not used [-Wunused-function]
  473 | static int spi_pl022_init(const struct device *dev)
      |            ^~~~~~~~~~~~~~
/home/crs/zephyrproject/zephyr/drivers/spi/spi_pl022.c:468:30: warning: 'spi_pl022_api' defined but not used [-Wunused-variable]
  468 | static struct spi_driver_api spi_pl022_api = {
      |                              ^~~~~~~~~~~~~
```

andrei-edward-popa replied:
Okay, so probably that's the right way.

yonsch commented on include/zephyr/dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h:
```diff
@@ -90,4 +90,25 @@
+#define SPI0_RX_P16 RP2040_PINMUX(16, RP2_PINCTRL_GPIO_FUNC_SPI)
+#define SPI0_CSN_P17 RP2040_PINMUX(17, RP2_PINCTRL_GPIO_FUNC_SPI)
+#define SPI0_SCK_P18 RP2040_PINMUX(18, RP2_PINCTRL_GPIO_FUNC_SPI)
+#define SPI0_TX_P19 RP2040_PINMUX(19, RP2_PINCTRL_GPIO_FUNC_SPI)
```
This should get up to pin 29, see table on page 13 of the RP2040 datasheet

soburi replied:
Fixed it.

yonsch reviewed (changes requested):
My review for the Pico parts

yonsch commented:
Two more things:
1. Why is CSN missing from the board's pinmux?
2. Set a default clock-frequency in `rpi_pico.dts`

soburi commented:
Hi, @yonsch,

> * Why is CSN missing from the board's pinmux?

It is my mistakes. Just forgot it.
CS pin is often controlled by 'cs-gpios' as gpio, so I misunderstood that I didn't must need it. 

> * Set a default clock-frequency in `rpi_pico.dts`

Set 8MHz to default. 
It can works also in peripheral mode (pclk/12 =~ 10MHz) and with enough safety margin.

yonsch commented on drivers/spi/spi_pl022.c:
```diff
@@ -0,0 +1,538 @@
+		/* Fill up fifo with available TX data */
+		while (SSP_TX_FIFO_NOT_FULL(cfg->reg) && tx_remaining > 0) {
+			/* Send 0 in the case of read operation */
+			uint32_t txd = 0;
```
Declare those variables at the beginning of the function

soburi replied:
Fixed it.

yonsch reviewed (changes requested):
Thanks, it looks better. I added a few comments about the PL022 driver, though it should be more thoroughly reviewed by someone with more expertise in this field.

soburi commented on drivers/spi/spi_pl022.c:
```diff
@@ -0,0 +1,538 @@
+static int spi_pl022_configure(const struct device *dev,
+			       const struct spi_config *spicfg)
+{
+	const struct spi_pl022_cfg *cfg = dev->config;
```
Fixed it.

soburi commented on boards/arm/rpi_pico/doc/index.rst:
```diff
@@ -60,6 +60,9 @@ hardware features:
      - :dtcompatible:`snps,designware-i2c`
+   * - SPI
+     - :kconfig:option:`CONFIG_SPI`
+     - :dtcompatible:`arm,pl022`
```
Fixed it.

soburi commented on drivers/spi/spi_pl022.c:
```diff
@@ -0,0 +1,538 @@
+
+#include "spi_context.h"
+
+#define GEN_MASK(len, pos) (((1 << len) - 1) << pos)
```
Fixed it.
Rewrite internally used defines.

soburi commented on include/zephyr/dt-bindings/pinctrl/rpi-pico-rp2040-pinctrl.h:
```diff
@@ -90,4 +90,35 @@
+#define SPI0_CSN_P21 RP2040_PINMUX(21, RP2_PINCTRL_GPIO_FUNC_SPI)
+#define SPI0_SCK_P22 RP2040_PINMUX(22, RP2_PINCTRL_GPIO_FUNC_SPI)
+#define SPI0_TX_P23 RP2040_PINMUX(23, RP2_PINCTRL_GPIO_FUNC_SPI)
+#define SPI0_RX_P24 RP2040_PINMUX(24, RP2_PINCTRL_GPIO_FUNC_SPI)
```
Oh.. It's my careless.
Fixed it.

yonsch reviewed (dismissed):
LGTM, Thanks!

yonsch commented:
@soburi To comply with #45410, please change includes to `#include <zephyr/..>` where applicable.

soburi commented:
@yonsch 

> @soburi To comply with #45410, please change includes to `#include <zephyr/..>` where applicable.

Thank you for notify.
I updated it.

mbolivar-nordic reviewed (commented):
Devicetree portions all LGTM. @tbursztyka could you please take a look at this one?

petejohanson commented:
@soburi Sadly, conflicts after the USB device driver info got added to the docs file. Needs a rebase.

petejohanson commented on boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi:
```diff
@@ -31,4 +31,24 @@
+		};
+	};
+
+	spi1_default: spi1_default {
```
According to the [Pico Datasheet](https://datasheets.raspberrypi.com/pico/Pico-R3-A4-Pinout.pdf) the only "default" SPI is SPI0, I would say we shouldn't enable/default `spi1` defaults at all for the pico board.

soburi replied:
Rebasing and Removed spi1 and related definitions.

Question: i2c1 is also not set as default in this illustration. Is it have no problem?
(I made SPI1 by seeing that.)

petejohanson replied:
I didn't have a chance to review the i2c PR, but I would suggest we remove that for consistency in a separate PR.

soburi replied:
Thanks for reply. I understand and agree it.

petejohanson commented on boards/arm/rpi_pico/rpi_pico.dts:
```diff
@@ -84,3 +86,17 @@
+	pinctrl-names = "default";
+};
+
+&spi1 {
```
Per my other comment, I think we shouldn't be setting up `spi1` at all with defaults, leave that to end users.

soburi replied:
ditto.

petejohanson reviewed (commented):
One minor comment on the spi1 defaults for the Pico board definition.

soburi committed:
tests: drivers: spi: loopback: add rpi_pico configuration

Add rpi_pico configuration for SPI loopback test.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi commented:
@petejohanson ,

> @soburi Sadly, conflicts after the USB device driver info got added to the docs file. Needs a rebase.

I rebased to newest.

tbursztyka commented on drivers/spi/spi_pl022.c:
```diff
@@ -0,0 +1,541 @@
+		spi_context_update_tx(ctx, 1, len);
+		spi_context_update_rx(ctx, 1, len);
+
+	} while (spi_pl022_transfer_ongoing(data));
```
Is there any possibility to get this driver working through interrupts instead of polling?

soburi replied:
> Is there any possibility to get this driver working through interrupts instead of polling?

Currently, only work with polling. Not handles interrupt.
But I think it should also support the CONFIG_SPI_ASYNC option by handling interrupts
I have a plan to implement an interrupts handler later.

soburi committed:
drivers: spi: pl022: Support interrupt based transfer

Add interrupt based transfer feature for the PL022 SPI.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

mbolivar-nordic reviewed (dismissed):
Thank you for the nice work on the devicetree and driver model parts, +1 for these.

yonsch commented:
@tbursztyka Could you give this another look?

yonsch commented:
@soburi Please rebase and fix conflicts

soburi commented:
@yonsch 

> @soburi Please rebase and fix conflicts

Thanks for notifying me.

I rebased it to the newest source.
I found a failure of tests/drivers/spi/spi_loopback test with the newest source.
And I fixed it and also updated this source.

carlescufi commented:
@petejohanson  and @andrei-edward-popa could you please take another look?

petejohanson reviewed (dismissed):
Tested here on a XIAO RP2040 controller on a keyboard w/ a shift register on the SPI bus, working perfectly w/ defaults, and also if I enable the interrupt mode.

Docs, DT bits, etc all look good. My high level understanding of the SPI driver pieces I don't see any major issues.

andrei-edward-popa commented:
Looks great to me! I cannot wait to try the SPI in a real application with Pico. Good job!

petejohanson reviewed (changes requested):
Code changes tested here, and working fine w/ my 595 shift register driver, and sharp memory display.

The compliance check is failing right now though, sadly, due to the commit message of the last commit being just "overlay".

soburi commented:
Hi, @petejohanson 
> Code changes tested here, and working fine w/ my 595 shift register driver, and sharp memory display.
> 
> The compliance check is failing right now though, sadly, due to the commit message of the last commit being just "overlay".

Very sorry. I accidentally pushed the code I put in for confirmation with it committed. Deleted the last commit.

mbolivar-nordic reviewed (dismissed):
DT bits look OK to me. @tbursztyka please take another look

yonsch commented:
@soburi Please check if this compiles when `CONFIG_LEGACY_INCLUDE_PATH=n`. If it doesn't, you should add a few `zephyr/` prefixes to your includes. One of these instances was fixed in a recent PR, so you will have to rebase first. Other than that, LGTM.

soburi commented:
@yonsch 
> @soburi Please check if this compiles when `CONFIG_LEGACY_INCLUDE_PATH=n`. If it doesn't, you should add a few `zephyr/` prefixes to your includes. One of these instances was fixed in a recent PR, so you will have to rebase first. Other than that, LGTM.

Thanks for notification.
I verified it, and found no problems.

soburi commented:
- Use DT_HAS_ARM_PL022_ENABLED instead of dt_compat_enabled in Kconfig.pl022
- Call spi_context_complete if SPI_ASYNC enabled even if not enabled SPI_PL022_INTERRUPT.

galak commented on dts/arm/rpi_pico/rp2040.dtsi:
```diff
@@ -101,6 +101,32 @@
+			resets = <&reset RPI_PICO_RESETS_RESET_SPI0>;
+			interrupts = <18 RPI_PICO_DEFAULT_IRQ_PRIORITY>;
+			interrupt-names = "spi0";
+			label = "SPI_0";
```
remove `label`,  not needed

galak commented on dts/arm/rpi_pico/rp2040.dtsi:
```diff
@@ -101,6 +101,32 @@
+			clocks = <&peripheral_clk>;
+			interrupts = <19 RPI_PICO_DEFAULT_IRQ_PRIORITY>;
+			interrupt-names = "spi1";
+			label = "SPI_1";
```
remove label

galak commented on tests/drivers/spi/spi_loopback/boards/rpi_pico.conf:
```diff
@@ -0,0 +1,12 @@
+#
+
+# enable rpi_pico spi instance for loopback test
+CONFIG_SPI_LOOPBACK_DRV_NAME="SPI_0"
```
Seen reworked PR for spi_loopback: https://github.com/zephyrproject-rtos/zephyr/pull/48222

soburi replied:
Hi, @galak, 

I'll prepare to apply the #48222 change to this PR.

I also failed #48222 same as @erwango observed.

@erwango mentioned in #48222

> @galak, as expected this test failed on our bench since #48358 was merged. Out of curiosity I launched a campaign using current PR and I still see issues so most likely it requires a bit more work. Do you have the same status on your side, or do you want me to dig a little bit more ?

I'm checking #48222 progress, and I'll report when the update is complete.

Thank you.

soburi replied:
Hi, @galak 

#48222 is correctly worked with also rpi_pico now.
I prepare to update this PR.

soburi replied:
Rebase done.

soburi commented:
Update done. This PR should merge after #48222 merged.

galak commented on dts/bindings/spi/arm,pl022.yaml:
```diff
@@ -0,0 +1,18 @@
+  clocks:
+    required: true
+
+  interrupts:
```
`interrupt` is `false` by default so not needed.

soburi replied:
Removed it.

soburi commented:
CI is unstable at this time. The fails seems not relate this PR.
I'll rebase if need.

yonsch commented:
@soburi Congratulations, and thank you for your patience!

soburi commented:
Hi, @yonsch ,

thank you for your helps.
And also please #45137, It makes life happy with GDB!
